### PR TITLE
Add IXMP_DATA / execute all tests in temporary directories

### LINKED ISSUE #124: Configuration & data are not in expected (XDG) locations
Several related issues here that could be fixed together.

1. `ixmp-config` places a file at `~/.local/ixmp/config.json`. This does not conform to the [XDG base directory specification](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html). The file should be at either:
    - `~/.local/share/ixmp/config.json` — i.e. using `$XDG_DATA_HOME` or its default.
    - `~/.config/ixmp.json` or `~/.config/ixmp/config.json` — i.e. using `$XDG_CONFIG_HOME` or its default.
2. Likewise the `localdb` directory created with `ixmp.Platform(dbtype='HSQLDB')` is incorrectly at `~/.local/ixmp/localdb` instead of `~/.local/share/ixmp/localdb`.
3. `ixmp.config.get()` raises an exception if the configuration file is missing. Sensible defaults should be used instead.

### PR DESCRIPTION
**Pull request by khaeru**
This code causes all tests to be executed in temporary directories, without touching the user's local database or configuration file. This is accomplished by checking a new environment variable `IXMP_DATA`, which points to a directory that is used for both `config.json` and the `DEFAULT_LOCAL_DB_PATH`. Along the way, handling of configuration keys is formalized a little.

To check that the code works as advertised, one can make one's own (user) ixmp data directory read-only, e.g. `$ chmod -w ~/.local/share/ixmp`. Then run the tests. The tests pass; this shows that `ixmp-config` calls in tests, and any local databases created/used, are not written in this directory.

Closes #124.

- New class `ixmp.config.Config` to regularize config handling.
- Remove `ixmp.default_paths`, `ixmp.default_path_constants` submodules.
- Scenario__init__(): use Config.find_dbprops() and Config.get().
- cli.config(): use Config.save().
- Update tests:
  - conftest.py: new fixture 'tmp_env' and hook 'pytest_sessionstart'.
  - test_config.py: test Config features.
  - test_core.py: use pytest 'caplog' fixture to test that configuration values are used.
  - test_tutorials.py: use 'tmp_env' to force tutorial execution in a sandbox.

**PR checklist:**
- [x] Tests added
- [x] Documentation added
- [x] Description in RELEASE_NOTES.md added

### CODE REVIEW

File: ixmp/config.py [hunk 1]
```diff
@@ -1,50 +1,208 @@
...
+            return
+        self.values[key] = value
+
+    def clear(self):
+        """Clear all configuration keys by setting their values to None."""
+        self.values = {key: None for key in self._keys}
+
+        # Set 'DEFAULT_LOCAL_DB_PATH'
+        # Use the first identifiable path
+        for label, config_dir in self._iter_paths():
```
Comment (gidden) [inline ixmp/config.py]: could this be simply:

```
label, config_dir = next(self._iter_paths())
self.values['DEFAULT_LOCAL_DB_PATH'] = (config_dir / 'localdb' / 'default')
```

File: ixmp/config.py [hunk 2]
```diff
@@ -1,50 +1,208 @@
...
+            break
+
+    def save(self):
+        """Write configuration keys to file.
+
+        `config.json` is created in the first of the ixmp configuration
+        directories that exists. Only non-null values are written.
+        """
+        # Use the first identifiable path
+        for label, config_dir in self._iter_paths():
```
Comment (gidden) [inline ixmp/config.py]: if above is true, then same here

File: ixmp/config.py (hunk 1 above)
Comment (khaeru) [reply ixmp/config.py]: Good eye. Yes. I'll make this change.

**Comment by gidden** [comment]
Hey @khaeru let me know if there's anything else to do here. Otherwise as soon as it passes, we can pull and continue with other PRs!

**Comment by khaeru** [comment]
Done, please go ahead and merge.

**Comment by gidden** [comment]
thanks @khaeru, great addition!

**Pull request merged**